MoreLinq/SelectAsync.cs

```diff
@@ -0,0 +1,439 @@
+        /// <para>
+        /// The <paramref name="selector"/> function should be designed to be
+        /// thread-agnostic.</para>
+        /// </remarks>
+
+        public static ISelectAsyncEnumerable<TResult> SelectAsync<T, TResult>(
+            this IEnumerable<T> source, Func<T, Task<TResult>> selector)
+        {
+            if (selector == null) throw new ArgumentNullException(nameof(selector));
+            return source.SelectAsync((e, _) => selector(e));
+        }
+
+        /// <summary>
+        /// Asynchronously projects each element of a sequence to its new form.
+        /// The projection function receives a <see cref="CancellationToken"/>
+        /// as an additional argument that can be used to abort any asynchronous
+        /// operations in flight.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// This method uses deferred execution semantics. The results are
+        /// yielded as each asynchronous projection completes and, by default,
+        /// not guaranteed to be based on the source sequence order. If order
+        /// is important, compose further using
+        /// <see cref="PreserveOrder{T}(MoreLinq.ISelectAsyncEnumerable{T})"/>
+        /// and a Boolean value of <c>true</c>.</para>
+        /// <para>
+        /// This method starts a new task where the asynchronous projections
+        /// are started and awaited.</para>
+        /// <para>
+        /// The <paramref name="selector"/> function should be designed to be
+        /// thread-agnostic.</para>
+        /// </remarks>
+
+        public static ISelectAsyncEnumerable<TResult> SelectAsync<T, TResult>(
+            this IEnumerable<T> source, Func<T, CancellationToken, Task<TResult>> selector)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (selector == null) throw new ArgumentNullException(nameof(selector));
+
+            return SelectAsyncEnumerable.Create(
+                options => _(options.MaxConcurrency ?? int.MaxValue,
+                             options.Scheduler ?? TaskScheduler.Default,
+                             options.PreserveOrder));
+
+            IEnumerable<TResult> _(int maxConcurrency, TaskScheduler scheduler, bool ordered)
+            {
+                var queue = new BlockingCollection<object>();
+                var cancellationTokenSource = new CancellationTokenSource();
+                var completed = false;
+
+                async Task<KeyValuePair<int, TResult>> Select(KeyValuePair<int, T> input, CancellationToken cancellationToken) =>
+                    new KeyValuePair<int, TResult>(input.Key, await selector(input.Value, cancellationToken).ConfigureAwait(false));
+
+                var item = source.Index().GetEnumerator();
+                IDisposable disposable = item; // disables AccessToDisposedClosure warnings
+                try
+                {
+                    Task.Factory.StartNew(async () =>
```

Comment:
Fair point.

Local methods can break the flow when reading top-down, forcing more on the cognitive stack. Moving into its own method means moving all (and the many) free variables into parameters, which when taken out of context may not read easily. The method has become large by supporting ordered results. I can give it a shot and see what it gives but I reckon that different tactics may yields better results.

Reply:
See how you feel about 33e437aa52582c11c497d041f1217332ed43401d. One bit of bad news is boxing, which results from being generic about the extract method and `BlockingCollection<>` generic type argument being `object`. I could offset that by introducing a discriminated union notification `struct` ([à la F#](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/discriminated-unions#struct-discriminated-unions)) but it seems like an overkill at this stage? Improvement for the future? The clarity (brought by the extraction) has come at some cost. Either we revert and improve by refactoring later or go forward and improve by optimising the boxing later. Both ways, speaking objectively, we end up at the same place. This is about internals, so like use of `List<>` not being the best data structure for insertion, this is something we should revisit later and focus instead on the shape of the API and the runtime behaviour for a first (experimental) release. What say?

Reply:
In 90acfbeff8f42729894da7c84a85447a96a76a88, I'm [using a tuple like you suggested some time ago](https://github.com/morelinq/MoreLINQ/pull/205#discussion_r85404147), except it's a value tuple so no heap allocation. The boxing concern is now also mute and there are fewer chances to make errors with casting decision (since we don't have tests) as the storage class for notifications is no longer `object`. I'm happy with the all the positive impacts resulting from the method extraction. Thanks for the challenge :+1: and hope you as well like where we've ended.